alicat/driver.py

```diff
@@ -0,0 +1,484 @@
+        if values[-1].upper() == 'LCK':
+            self.button_lock = True
+            del values[-1]
+        else:
+            self.button_lock = False
+        if len(values) == 5 and len(self.keys) == 6:
+            del self.keys[-2]
+        elif len(values) == 7 and len(self.keys) == 6:
+            self.keys.insert(5, 'total flow')
+        elif len(values) == 2 and len(self.keys) == 6:
+            self.keys.insert(1, 'setpoint')
+        return {k: (float(v) if self._is_float(v) else v)
+                for k, v in zip(self.keys, values)}
+
+    async def set_gas(self, gas):
+        """Set the gas type.
+
+        Args:
+            gas: The gas type, as a string or integer. Supported strings are:
+                'Air', 'Ar', 'CH4', 'CO', 'CO2', 'C2H6', 'H2', 'He', 'N2',
+                'N2O', 'Ne', 'O2', 'C3H8', 'n-C4H10', 'C2H2', 'C2H4',
+                'i-C2H10', 'Kr', 'Xe', 'SF6', 'C-25', 'C-10', 'C-8', 'C-2',
+                'C-75', 'A-75', 'A-25', 'A1025', 'Star29', 'P-5'
+
+                Gas mixes may only be called by their mix number.
+        """
+        self._test_controller_open()
+
+        if isinstance(gas, str):
+            if gas not in self.gases:
+                raise ValueError(f"{gas} not supported!")
+            gas_number = self.gases.index(gas)
+        else:
+            gas_number = gas
+        command = f'{self.unit}$$W46={gas_number}'
+        await self._write_and_read(command)
+        reg46 = await self._write_and_read('f{self.unit}$$R46')
+        reg46_gasbit = int(reg46.split()[-1]) & 0b0000000111111111
+
+        if gas_number != reg46_gasbit:
+            raise OSError("Cannot set gas.")
+
+    async def create_mix(self, mix_no, name, gases) -> None:
+        """Create a gas mix.
+
+        Gas mixes are made using COMPOSER software.
+        COMPOSER mixes are only allowed for firmware 5v or greater.
+
+        Args:
+        mix_no: The mix number. Gas mixes are stored in slots 236-255.
+        name: A name for the gas that will appear on the front panel.
+        Names greater than six letters will be cut off.
+        gases: A dictionary of the gas by name along with the associated
+        percentage in the mix.
+        """
+        firmware = await self.get_firmware()
+        if any(v in firmware for v in ['2v', '3v', '4v', 'GP']):
+            raise OSError("This unit does not support COMPOSER gas mixes.")
+
+        if mix_no < 236 or mix_no > 255:
+            raise ValueError("Mix number must be between 236-255!")
+
+        total_percent = sum(gases.values())
+        if total_percent != 100:
+            raise ValueError("Percentages of gas mix must add to 100%!")
+
+        if any(gas not in self.gases for gas in gases):
+            raise ValueError("Gas not supported!")
+
+        gas_list = ' '.join(
+            [' '.join([str(percent), str(self.gases.index(gas))])
+                for gas, percent in gases.items()])
```

Comment:
Not very readable. Maybe...

```python

gas_list = [f'{percent} {self.gases.index(gas)}' for gas, percent in gases.items()]
command = ' '.join([self.unit,
                    'GM',
                    name,
                    str(mix_no),
                    ' '.join(gas_list)])
```

Reply:
b6a37900b07e3e5434bb88be8cc7788ff1fce1e3

Reply:
Reminder to me: test again on real hardware